Update CONTRIBUTING.md with workflow instructions (#2118)
dotnet/dotnet-docker: .github/CONTRIBUTING.md

> # Contributing

Author: @mthalman - thoughts on moving this to the root of the repo to give it more visibility?  I suggest this because the repo's README.md doesn't reference it like most repos do.

Reviewer: You mean like adding a section like [this](https://github.com/dotnet/runtime#how-can-i-contribute) that links to this file?

Author: I am not proposing any changes to the READMEs.  I don't think the contribution guide should be mentioned in the Docker Hub readmes.  I am simply suggesting we move CONTRIBUTING.md to the root of the repo to give it more visibility.  The CONTRIBUTING.md file can be located at either the root or within the .github folder. FWIW, the file is located at the root of https://github.com/dotnet/runtime.

Reviewer: Ok, I think moving the file is a good idea.

See [dotnet/runtime Contributing](https://github.com/dotnet/runtime/blob/master/CONTRIBUTING.md) for information about coding styles, source structure, making pull requests, and more.

## Branches

When making PRs, all source code changes (e.g. Dockerfiles, tests, and infrastructure) should be made in the [nightly branch](https://github.com/dotnet/dotnet-docker/tree/nightly). Only changes to the samples and documentation will be accepted against the [master branch](https://github.com/dotnet/dotnet-docker/tree/master).

## Workflow Instructions

### Building

The [`build-and-test.ps1`](https://github.com/dotnet/dotnet-docker/blob/master/build-and-test.ps1) script will build and test the .NET Docker images. Given the matrix of supported .NET versions, distros, and architectures there are numerous Dockerfiles and building can take a while. To make this manageable, the script supports several options for filtering down what images get built and tested.

> - Build and test all of the .NET 5.0 images for the current platform (e.g. linux/x64, linux/arm, linux/arm64, windows/x64).

Reviewer: ```suggestion
- Build and test all of the .NET 5.0 images for the Docker platform your machine is targeting (e.g. linux/x64, linux/arm, linux/arm64, windows/x64).
```

    ``` console
    > ./build-and-test.ps1 -Version 5.0
    ```

- Build the 3.1 Nano Server 1809 images

    ``` console
    > ./build-and-test.ps1 -Version 3.1 -OS nanoserver-1809 -Mode Build
    ```

- Build and test the samples

    ``` console

>     > ./build-and-test.ps1 -Path *samples*

Reviewer: ```suggestion
    > ./build-and-test.ps1 -Path *samples* -TestCategories @("sample")
```

Author: I'm changing this to `> ./build-and-test.ps1 -Path *samples* -TestCategories sample` instead as this is the shorthand way that PS autocomplete creates for you.

    ```

- Test the 5.0 Alpine 3.12 images for the current architecture (e.g. x64, arm, arm64).

    ``` console
    > ./build-and-test.ps1 -Version 5.0 -OS alpine3.12 -Mode Test
    ```

### Editing Dockerfiles

> The [Dockerfiles](https://github.com/search?q=repo%3Adotnet%2Fdotnet-docker+filename%3ADockerfile&type=Code&ref=advsearch&l=&l=) checked into this repo are generated from a set of [templates](https://github.com/dotnet/dotnet-docker/tree/master/eng/dockerfile-templates). A single templates generate the set of Dockerfiles that are similar (e.g. all Windows sdk Dockerfiles for a particular .NET version).  This ensure consistency across the various Dockerfiles and eases the burden of making changes to the Dockerfiles.  Instead of editing the Dockerfiles directly, the templates should be updated and then the Dockerfiles should get regenerated by running the [generate Dockerfiles script](https://github.com/dotnet/dotnet-docker/blob/master/eng/dockerfile-templates/Get-GeneratedDockerfiles.ps1)

Reviewer: ```suggestion
The [Dockerfiles](https://github.com/search?q=repo%3Adotnet%2Fdotnet-docker+filename%3ADockerfile&type=Code&ref=advsearch&l=&l=) contained in this repo are generated from a set of [templates](https://github.com/dotnet/dotnet-docker/tree/master/eng/dockerfile-templates). A single template generates the set of Dockerfiles that are similar (e.g. all Windows sdk Dockerfiles for a particular .NET version).  This ensures consistency across the various Dockerfiles and eases the burden of making changes to the Dockerfiles.  Instead of editing the Dockerfiles directly, the templates should be updated and then the Dockerfiles should get regenerated by running the [generate Dockerfiles script](https://github.com/dotnet/dotnet-docker/blob/master/eng/dockerfile-templates/Get-GeneratedDockerfiles.ps1)
```

### Tests

There are two basic types of [tests](https://github.com/dotnet/dotnet-docker/tree/master/tests) for each of the images produced from this repo.

1. Unit tests that validate the static state of images.  This includes things like verifing which environment variables are defined.
1. Scenario tests that validate basic usage scenarios.  For example the SDK image is used to create, build and run a new console app.

When editing Dockerfiles, please ensure the appropriate test changes are also made.

### Metadata Changes

The [`manifest.json`](https://github.com/dotnet/dotnet-docker/blob/master/manifest.json) contains metadata used by the engineering infrastructure to build and publish the images.  It includes information such as:

- Dockerfiles to build
- Image Tags
- Manifest/shared tags to create and which images they reference
- Docker repositories to publish the images to

> - Dockerfile templates used to generated the Dockerfiles

Reviewer: ```suggestion
- Dockerfile templates used to generate the Dockerfiles
```
